File: data-jdbc-jooq/src/main/java/com/example/interfaces/PostController.java
package com.example.interfaces;

import com.example.application.CreatePostCommand;
import com.example.application.PaginatedResult;
import com.example.application.PostCommandService;
import com.example.application.PostQueryService;
import com.example.application.PostSummary;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.QueryValue;
import lombok.RequiredArgsConstructor;

import java.net.URI;

@Controller
@RequiredArgsConstructor
public class PostController {
    private final PostQueryService postQueryService;
    private final PostCommandService postCommandService;

    @Get("/posts")
    public PaginatedResult<PostSummary> findPosts(
            @QueryValue(value = "q", defaultValue = "") String keyword,
            @QueryValue(defaultValue = "0") int offset,
            @QueryValue(defaultValue = "10") int limit
    ) {
        return postQueryService.findPostsByKeyword(keyword, offset, limit);
    }

    @Post("/posts")
    public HttpResponse<?> createPost(@Body CreatePostCommand command) {
        var savedPostID = postCommandService.createPost(command);
        return HttpResponse.created(URI.create("/posts/" + savedPostID));
    }
}

File: data-r2dbc-kotlin-co/src/main/kotlin/com/example/PostController.kt
package com.example

import io.micronaut.http.HttpResponse
import io.micronaut.http.HttpResponse.*
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.validation.Validated
import kotlinx.coroutines.flow.Flow
import java.net.URI
import java.util.*
import jakarta.validation.Valid

@Controller("/posts")
@Validated
class PostController(private val posts: PostRepository) {

    @Get(uri = "/", produces = [MediaType.APPLICATION_JSON])
    fun all(): HttpResponse<Flow<Post>> = ok(posts.findAll())

    @Get(uri = "/{id}", produces = [MediaType.APPLICATION_JSON])
    suspend fun byId(@PathVariable id: UUID): HttpResponse<Any> {
        val post = posts.findById(id) ?: return notFound()
        return ok(post)
    }

    @io.micronaut.http.annotation.Post(consumes = [MediaType.APPLICATION_JSON])
    suspend fun create(@Body @Valid body: CreatePostCommand): HttpResponse<Any> {
        val data = Post(title = body.title, content = body.content)
        val saved = posts.save(data)
        return created(URI.create("/posts/" + saved.id))
    }
}